### PR TITLE
Add inline variable refactoring

### LINKED ISSUE #2507: Add Inline temp variable code action
### Description

Replace a temporary variable with its value at all usage sites.

## Example

**Before:**

```swift
func example() {
    let basePrice = item.price
    let total = basePrice * quantity
}
```

**After:**

```swift
func example() {
    let total = item.price * quantity
}
```

### PR DESCRIPTION
**Pull request by somiljain2006**
Fixes #2507 

Implements Inline Variable for local let declarations with lexical lookup

### CODE REVIEW

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+extension SyntaxProtocol {
+  /// Traverses up the syntax tree to find the first ancestor of the specified type.
+  func firstMatch<T: SyntaxProtocol>(ofType type: T.Type) -> T? {
+    var current: Syntax? = Syntax(self)
+    while let node = current {
+      if let match = node.as(T.self) { return match }
+      current = node.parent
+    }
+    return nil
+  }
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: We should have `ancestorOrSelf` for this already.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+      let range = variableDecl.positionAfterSkippingLeadingTrivia..<variableDecl.endPosition
+      edits.append(SourceEdit(range: range, replacement: newDecl.description))
+    }
+
+    return edits
+  }
+
+  private static func findValidUsages(
+    targetIdentifier: String,
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Instead of using `String` here, we should base all of this on top of `Identifier`, which should also make sure that we match `let foo = 2` with a usage that spells the variable name in backticks like ``print(`foo`)``.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    targetBinding: PatternBindingSyntax,
+    in scope: CodeBlockItemListSyntax
+  ) -> [DeclReferenceExprSyntax] {
+    let finder = ReferenceCollector(targetIdentifier: targetIdentifier)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Instead of a two-part process, could `ReferenceCollector` directly use SwiftLexicalLookup to check if the reference refers to the variable in question?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    return finder.candidates.filter { usageNode in
+      guard usageNode.position > targetBinding.endPosition else { return false }
+
+      let lookupResults = usageNode.lookup(Identifier(usageNode.baseName))
+
+      for result in lookupResults {
+        switch result {
+        case .fromScope(_, let names):
+          let matchesTarget = names.contains { name in
+            if let resolvedBinding = name.syntax.firstMatch(ofType: PatternBindingSyntax.self) {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Do we need to walk up the tree to a `PatternBindingSyntax` here or could we compare `name.syntax` directly? I’m wondering because walking up the tree could theoretically find a completely unrelated `PatternBindingSyntax`.

Same for the check below of `VariableDeclSyntax`.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        }
+      }
+      return false
+    }
+  }
+
+  /// Determines whether it is safe to inline the initializer at multiple usage sites.
+  /// If the initializer has side effects (like a function call), duplicating it could
+  /// change the program's behavior.
+  private static func canDuplicateInitializer(
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Moving the location at which the initializer is called may already change program behavior. If the user invokes the refactoring action, they are opting into moving the initializer around (and potentially duplicating them), so I don’t think we should try to be too clever here. There may be good reasons to inline variables initialized by a function call because the developer knows that it’s save to duplicate.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    return initializer.description
+  }
+
+  static func isPrimaryExpression(_ expr: ExprSyntax) -> Bool {
+    expr.is(IntegerLiteralExprSyntax.self) || expr.is(FloatLiteralExprSyntax.self)
+      || expr.is(BooleanLiteralExprSyntax.self) || expr.is(StringLiteralExprSyntax.self)
+      || expr.is(NilLiteralExprSyntax.self) || expr.is(DeclReferenceExprSyntax.self)
+      || expr.is(FunctionCallExprSyntax.self) || expr.is(MemberAccessExprSyntax.self)
+      || expr.is(SubscriptCallExprSyntax.self) || expr.is(ArrayExprSyntax.self) || expr.is(DictionaryExprSyntax.self)
+      || expr.is(TupleExprSyntax.self) || expr.is(ClosureExprSyntax.self)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: I think this would read a little nicer as a switch over `expr.as(ExprSyntaxEnum.self)`.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    guard let parent = usageNode.parent else { return false }
+
+    if parent.is(MemberAccessExprSyntax.self) || parent.is(SubscriptCallExprSyntax.self)
+      || parent.is(FunctionCallExprSyntax.self) || parent.is(PrefixOperatorExprSyntax.self)
+      || parent.is(PostfixOperatorExprSyntax.self)
+    {
+      return !isPrimaryExpression(initializer)
+    }
+
+    let parentSeq = parent.as(SequenceExprSyntax.self) ?? parent.parent?.as(SequenceExprSyntax.self)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Why do we need to check the parent and grandparent but not great-grandparent?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+
+    guard let parentSeq = parentSeq else {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]:
```suggestion
    guard let parentSeq else {
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+      return !isPrimaryExpression(initializer)
+    }
+
+    var newElements: [ExprSyntax] = []
+    var replaced = false
+
+    for element in parentSeq.elements {
+      if element.id == usageNode.id {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: This doesn’t work if `parentSeq = parent.parent?.as(SequenceExprSyntax.self)`, right?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+      || expr.is(TupleExprSyntax.self) || expr.is(ClosureExprSyntax.self)
+  }
+
+  // Determine whether replacing the reference with the initializer would
+  // change the expression's operator grouping.
+  static func needsParentheses(
+    initializer: ExprSyntax,
+    usageNode: DeclReferenceExprSyntax,
+    operatorTable: OperatorTable
+  ) -> Bool {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift:155]: There are also other places in which parentheses need to be added outside of operator precedence. For example

```swift
let x = try three()
let result = 1 + x
```

but the following is invalid


```swift
let result = 1 + try three()
```

What do you think about the following approach instead: You parse the rewritten source code without parentheses and locate the node at the replaced offset. You then check if that node has the exact same text representation as the initializer that was inserted. If the rewritten text doesn’t parse or you don’t find the node, try again with parentheses.

For example if you replace `2 * x` by `2 * 1 + 1`, then you won’t be able to find any node in the syntax tree that represents `1 + 1`, so you know that parentheses need to be added.

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 11]
```diff
@@ -4027,6 +4027,625 @@ final class CodeActionTests: SourceKitLSPTestCase {
     )
   }
 
+  func testInlineVariableBasic() async throws {
+    try await assertCodeActions(
+      """
+      func doSomething() {1️⃣
+        let 0️⃣x = 422️⃣
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift]: We generally use `1️⃣` as the first location marker instead of `0️⃣`. Could you change these tests to be consistent with that pattern?

**Review by ahoppen — COMMENTED** [review]
Thanks for tackling this, @somiljain2006. I left a few suggestions inline.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+/// A code action that replaces a local variable with its assigned value at all usage sites.
+struct InlineVariable: SyntaxRefactoringCodeActionProvider {
+  package static let title: String = "Inline variable"
+
+  package static func nodeToRefactor(in scope: SyntaxCodeActionScope) -> PatternBindingSyntax? {
+    guard let node = scope.innermostNodeContainingRange else { return nil }
+
+    let variableDecl =
+      node.ancestorOrSelf(mapping: { $0.as(VariableDeclSyntax.self) })
+      ?? node.as(CodeBlockItemSyntax.self)?.item.as(VariableDeclSyntax.self)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Wouldn’t `node.as(CodeBlockItemSyntax.self)?.item.as(VariableDeclSyntax.self)` already imply that `node.ancestorOrSelf(mapping: { $0.as(VariableDeclSyntax.self) })` returned non-nil?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+/// A code action that replaces a local variable with its assigned value at all usage sites.
+struct InlineVariable: SyntaxRefactoringCodeActionProvider {
+  package static let title: String = "Inline variable"
+
+  package static func nodeToRefactor(in scope: SyntaxCodeActionScope) -> PatternBindingSyntax? {
+    guard let node = scope.innermostNodeContainingRange else { return nil }
+
+    let variableDecl =
+      node.ancestorOrSelf(mapping: { $0.as(VariableDeclSyntax.self) })
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: I think we should add `stoppingAt` for `CodeBlockSyntax` and `MemberBlockSyntax` so that we don’t offer this eg. in

```swift
let foo = {
  compute^HERE^Something()
}
```


```suggestion
      node.findParentOfSelf(
      ofType: VariableDeclSyntax.self,
      stoppingIf: { $0.is(CodeBlockSyntax.self) || $0.is(MemberBlockSyntax.self) }
    )
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    guard let variableDecl = variableDecl,
+      variableDecl.bindingSpecifier.tokenKind == .keyword(.let)
+    else {
+      return nil
+    }
+
+    // If the cursor is on the identifier or value, return the specific pattern binding.
+    if let specificBinding = node.ancestorOrSelf(mapping: { $0.as(PatternBindingSyntax.self) }) {
+      return specificBinding
+    }
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Since we are walking the tree two separate times, I am a little cautious in case `specificBinding` would resolve to a different binding than the variable decl. I think it would happen in this case

```swift
var x = {
  ^HERE^let value = 2
  return value
}()
```

Could we do something like this instead:
1. First find a `PatternBindingDecl`
2. If that doesn’t exist find a `VariableDeclSyntax` and then pick the first pattern (this is so we can offer the refactoring action on `let`)
3. Now walk up from the `PatternBindingDecl` using `.parent.parent.parent` or something like that to fetch the corresponding `VariableDeclSyntax` and check that it is `let`.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+      let initializer = binding.initializer?.value
+    else {
+      throw RefactoringNotApplicableError("Could not extract variable information")
+    }
+
+    guard let targetIdentifier = Identifier(identifierPattern.identifier) else {
+      throw RefactoringNotApplicableError("Could not parse the target identifier")
+    }
+
+    guard let blockScope = variableDecl.ancestorOrSelf(mapping: { $0.as(CodeBlockItemListSyntax.self) }) else {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Instead of walking up the entire tree, could you just check if the grandparent  is a `CodeBlockItemListSyntax`. I just want to make sure that we don’t extract the incorrect item eg. in this case:


```swift
func foo() {
  struct Foo {
    let x^HERE^ = "hi"
  }
  print("wohoo")
}
```

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 16]
```diff
@@ -4027,6 +4027,389 @@ final class CodeActionTests: SourceKitLSPTestCase {
...
+                TextEdit(
+                  range: positions["2️⃣"]..<positions["3️⃣"],
+                  newText: ""
+                ),
+              ]
+            ]
+          )
+        )
+      ]
+    }
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift:4371]: Could you also add a test case like the following that checks we match identifiers with backticks:

```swift
func foo() {
  let `x` = "hi"
  print(x)
  print(`x`)
}
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+  init(targetIdentifier: Identifier, targetBinding: PatternBindingSyntax) {
+    self.targetIdentifier = targetIdentifier
+    self.targetBinding = targetBinding
+    super.init(viewMode: .sourceAccurate)
+  }
+
+  override func visit(_ node: DeclReferenceExprSyntax) -> SyntaxVisitorContinueKind {
+    guard let nodeIdentifier = Identifier(node.baseName),
+      nodeIdentifier == targetIdentifier,
+      node.position > targetBinding.endPosition
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Technically, this lexical check isn’t necessary or correct. For example, this compiles in top-level code (albeit, it mis-compiles – known issue – but still):

```swift
print(x)
let x = 1
```

If you want to have some fun, try to guess what it prints before executing it.

I would just trust `SwiftLexicalLookup` for these checks.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+    else {
+      return .visitChildren
+    }
+
+    let lookupResults = node.lookup(nodeIdentifier)
+    for result in lookupResults {
+      switch result {
+      case .fromScope(_, let names):
+        let matchesTarget = names.contains { name in
+          name.syntax.id == targetBinding.id || name.syntax.id == targetBinding.pattern.id
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Why do we need to check both `targetBinding.id` and `targetBinding.pattern.id`?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        if matchesTarget {
+          candidates.append(node)
+          return .visitChildren
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: I think you can avoid this and just use the `return` two lines below.

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 20]
```diff
@@ -4027,6 +4027,389 @@ final class CodeActionTests: SourceKitLSPTestCase {
     )
   }
 
+  func testInlineVariableBasic() async throws {
+    try await assertCodeActions(
+      """
+      func doSomething() {2️⃣
+        let 1️⃣x = 423️⃣
+        print(4️⃣x5️⃣)
+      }
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift]: Nitpick: Could you order the position markers in-order inside the source code so that you invoke the refactoring action on `2`?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+    finder.walk(scope)
+    return finder.candidates
+  }
+
+  static func formatReplacement(
+    initializer: ExprSyntax,
+    usageNode: DeclReferenceExprSyntax
+  ) -> String {
+    if needsParentheses(initializer: initializer, usageNode: usageNode) {
+      return "(\(initializer.description.trimmingCharacters(in: .whitespacesAndNewlines)))"
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Instead of trimming the newlines, could you create a `TupleExprSyntax` and just transfer the leading and trailing trivia from `initializer` over to it?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    return initializer.description
+  }
+
+  static func needsParentheses(
+    initializer: ExprSyntax,
+    usageNode: DeclReferenceExprSyntax
+  ) -> Bool {
+    let root = usageNode.root
+    let fullText = root.description
+    let initializerText = initializer.description.trimmingCharacters(in: .whitespacesAndNewlines)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: I think this won’t work if you have eg. 

```swift
let foo = /* comment */
  1 * 2
```

Instead of trimming characters from the string, could you inspect the trivia pieces instead? Or maybe you don’t need to trim the trivia at all.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+    let startPos = usageNode.positionAfterSkippingLeadingTrivia.utf8Offset
+    let endPos = usageNode.endPositionBeforeTrailingTrivia.utf8Offset
+
+    let startIdx = fullText.utf8.index(fullText.utf8.startIndex, offsetBy: startPos)
+    let endIdx = fullText.utf8.index(fullText.utf8.startIndex, offsetBy: endPos)
+
+    var rewrittenText = fullText
+    rewrittenText.replaceSubrange(startIdx..<endIdx, with: initializerText)
+
+    let parsedFile = Parser.parse(source: rewrittenText)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: I don’t think we need to parse the full source file. You could try and find the surrounding `CodeBlockItemSyntax` and then just re-parse that. This should be more efficient for large source files.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+    let nodeStart = node.positionAfterSkippingLeadingTrivia.utf8Offset
+    let nodeEnd = node.endPositionBeforeTrailingTrivia.utf8Offset
+
+    if nodeStart == span.lowerBound && nodeEnd == span.upperBound {
+      if node.description.trimmingCharacters(in: .whitespacesAndNewlines) == text {
+        return true
+      }
+    }
+
+    for child in node.children(viewMode: .sourceAccurate) {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: You’ve effectively re-implemented a `SyntaxVisitor` here. I think it would be simpler to implement this as a `SyntaxAnyVisitor` instead.

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 25]
```diff
@@ -4027,6 +4027,389 @@ final class CodeActionTests: SourceKitLSPTestCase {
...
+                TextEdit(range: positions["2️⃣"]..<positions["3️⃣"], newText: ""),
+              ]
+            ]
+          )
+        )
+      ]
+    }
+  }
+
+  func testInlineVariableIgnoresShadowedReferences() async throws {
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift]: Isn’t this effectively already covered by `testInlineVariableRespectsShadowing`?

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 26]
```diff
@@ -4027,6 +4027,389 @@ final class CodeActionTests: SourceKitLSPTestCase {
...
+            let x = 100
+            print(x)
+          }
+        }
+        """,
+      atMarker: "1️⃣"
+    )
+  }
+
+  func testInlineVariableRespectsShadowingBetweenValidReferences() async throws {
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift]: Isn’t this effectively already covered by `testInlineVariableRespectsShadowing`?

File: Tests/SourceKitLSPTests/CodeActionTests.swift [hunk 27]
```diff
@@ -4027,6 +4027,389 @@ final class CodeActionTests: SourceKitLSPTestCase {
...
+                ),
+              ]
+            ]
+          )
+        )
+      ]
+    }
+  }
+
+  func testInlineVariableShadowingStartsAfterOuterReference() async throws {
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CodeActionTests.swift]: I don’t think this adds too much value. You’re just testing one edge case of `SwiftLexicalLookup` here, which should be covered by the `SwiftLexicalLookup` tests.

**Review by ahoppen — COMMENTED** [review]
Thanks! Looks a lot simpler now already. I have left a few more suggestions inline.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+      return parenthesized.description
+    }
+    return initializer.description
+  }
+
+  static func needsParentheses(
+    initializer: ExprSyntax,
+    usageNode: DeclReferenceExprSyntax
+  ) -> Bool {
+    guard let codeBlockItem = containingCodeBlockItem(for: Syntax(usageNode)) else {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Can’t this just be 
```suggestion
    guard let codeBlockItem = usageNode.ancestorOrSelf(mapping: { $0.as(CodeBlockItemSyntax.self) }) else {
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+      return true
+    }
+    let scopeText = codeBlockItem.description
+    let initializerText = initializer.trimmedDescription
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: In `formatReplacement`, we return `initializer.description`, so we should also be using `initializer.description` here instead of `trimmedDescription`.

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+      scopeText.utf8.startIndex,
+      offsetBy: startPos
+    )
+    let endIdx = scopeText.utf8.index(
+      scopeText.utf8.startIndex,
+      offsetBy: endPos
+    )
+    var rewrittenText = scopeText
+    rewrittenText.replaceSubrange(startIdx..<endIdx, with: initializerText)
+    let parsedScope = Parser.parse(source: rewrittenText)
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Let’s explicitly parse this as a `CodeBlockItemSyntax` here instead of building a full `SourceFileSyntax`.

```suggestion
    var parser = Parser(rewrittenText)
    let parsedScope = CodeBlockItemSyntax.parse(from: &parser)
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+
+  static func containsNode(
+    in node: Syntax,
+    withSpan span: Range<Int>,
+    matching text: String
+  ) -> Bool {
+    let finder = TargetNodeFinder(targetSpan: span, targetText: text)
+    finder.walk(node)
+    return finder.found
+  }
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Just a thought: Would this read nicer as a static method on `TargetNodeFinder` (also direct you sooner to the type that actually implements it)

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 32]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+/// A visitor that checks if a syntax node with a matching span and description exists.
+private final class TargetNodeFinder: SyntaxAnyVisitor {
+  let targetSpan: Range<Int>
+  let targetText: String
+  var found = false
+
+  init(targetSpan: Range<Int>, targetText: String) {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Instead of `Range<Int>`, can you use `Range<AbsolutePosition>` to indicate that these are UTF-8 positions?

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 33]
```diff
@@ -0,0 +1,286 @@
...
+
+  override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+    if found {
+      return .skipChildren
+    }
+
+    // Prune subtree traversal if full span does not overlap target range
+    let fullStart = node.position.utf8Offset
+    let fullEnd = node.endPosition.utf8Offset
+    if fullStart > targetSpan.upperBound || fullEnd < targetSpan.lowerBound {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Could this be simplified to the following (also, I think `contains` is slightly more restrictive and also sufficient)

```suggestion
    guard node.range.contains(targetSpan) else {
```

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 34]
```diff
@@ -0,0 +1,286 @@
...
+      return .skipChildren
+    }
+
+    let nodeStart = node.positionAfterSkippingLeadingTrivia.utf8Offset
+    let nodeEnd = node.endPositionBeforeTrailingTrivia.utf8Offset
+    if nodeStart == targetSpan.lowerBound && nodeEnd == targetSpan.upperBound {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: This should also be simpler if you use `Range<AbsolutePosition>`

**Review by ahoppen — COMMENTED** [review]
I have a couple more nitpicky comments comments, otherwise looks great to me 👍🏽

File: Sources/SwiftSyntaxCodeActions/InlineVariable.swift [hunk 35]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    // Prune subtree traversal if the node does not contain the target span.
+    guard node.range.contains(targetSpan) else {
+      return .skipChildren
+    }
+
+    let nodeStart = node.positionAfterSkippingLeadingTrivia
+    let nodeEnd = node.endPositionBeforeTrailingTrivia
+    if nodeStart == targetSpan.lowerBound && nodeEnd == targetSpan.upperBound {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxCodeActions/InlineVariable.swift]: Isn’t this equivalent to

```suggestion
    if node.trimmedRange == targetSpan {
```

**Review by ahoppen — COMMENTED** [review]
One last nitpick, then we’re ready to ship.

**Review by ahoppen — APPROVED** [review]
Let’s 🚢 it

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by somiljain2006** [comment]
@ahoppen I'm having trouble identifying the exact cause of the Windows CI failure. The build fails during CMake generation because the SwiftSyntax::SwiftLexicalLookup target cannot be found. I checked the top-level CMakeLists.txt, where SwiftSyntax is already brought in with find_package(SwiftSyntax CONFIG REQUIRED), but I haven't been able to figure out whether the issue is with how the dependency is declared in SourceKit-LSP or with the SwiftSyntax CMake package used by the Windows Stage2 toolchain. Can you tell me what I'm missing?

**Comment by ahoppen** [comment]
Seems like we need to expose `SwiftLexicalLookup` before we can use it because this is the first change that’s using it. Opened https://github.com/swiftlang/swift/pull/91985 to do so.

**Comment by ahoppen** [comment]
The Windows issue should be fixed now. Let's try again. 

@swift-ci Please test Windows

**Comment by somiljain2006** [comment]
@ahoppen How about extending the inline variable code action to support var declarations?

**Comment by ahoppen** [comment]
I don’t think extending it to `var` is all that important and reasoning about whether the variable is never modified is non-trivial (eg. need to also consider if it’s ever passed as inout to a function). But if you have a good proposal of how to implement it, I won’t object.

**Comment by somiljain2006** [comment]
@ahoppen I was thinking of this approach: only offer the action when the variable has an initializer and all resolved references are read-only. I would reject direct mutations (=, +=, etc.), inout uses such as &x, and any cases where a known mutation pattern is detected. More complex cases like closure captures and mutable pointer APIs could remain unsupported initially.

**Comment by ahoppen** [comment]
I think `+=` actually raises a good point where you can’t fully decide if a variable is read-only. You could define your own operator `++=` that takes a first argument as `inout`. Then, if you write, `x ++= [2]`, `x` would be mutated but you had no way of knowing that without running the type checker. 

I personally think that this would be a fairly big investment to do correctly with marginal pay-offs, so I wouldn’t do it.
